resources/sugarcrm/sugarsoapprotocol.cpp

```diff
@@ -0,0 +1,89 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "sugarsoapprotocol.h"
+#include "sugarjob.h"
+#include "sugarsession.h"
+#include <QNetworkReply>
+#include <KLocalizedString>
+#include <KDebug>
+
+SugarSoapProtocol::SugarSoapProtocol()
+{
+}
+
+int SugarSoapProtocol::login(const QString &user, const QString &password, QString &sessionId, QString &errorMessage)
+{
+    Q_ASSERT(mSession->soap() != nullptr);
+
+    const QByteArray passwordHash = password.toUtf8();
+
+    KDSoapGenerated::TNS__User_auth userAuth;
+    userAuth.setUser_name(user);
+    userAuth.setPassword(QString::fromAscii(passwordHash));
+    userAuth.setVersion(QLatin1String(".01"));
+
+    KDSoapGenerated::TNS__Set_entry_result entry_result = mSession->soap()->login(userAuth, QLatin1String("FatCRM"));
+    if (entry_result.error().number() == "0") {
+        sessionId = entry_result.id();
+        return KJob::NoError;
+    } else {
+        errorMessage = i18nc("@info:status", "Login for user %1 on %2 failed: %3", user, mSession->host(), mSession->soap()->lastError());
+        int faultcode = mSession->soap()->lastErrorCode();
+        if (faultcode == QNetworkReply::UnknownNetworkError ||
+           faultcode == QNetworkReply::HostNotFoundError) {
+            return SugarJob::CouldNotConnectError;
+        } else {
+            return SugarJob::LoginError;
+        }
+    }
+}
+
+void SugarSoapProtocol::logout()
+{
+    if (mSession->sessionId().isEmpty() && mSession->soap() != nullptr) {
+        KDSoapGenerated::TNS__Error_value errorValue = mSession->soap()->logout(mSession->sessionId());
+        if (errorValue.number() != "0") {
+            kDebug() << "logout returned error" << errorValue.number() << errorValue.name() << errorValue.description();
+        }
+        if (!mSession->soap()->lastError().isEmpty()) {
+            kDebug() << "logout had fault" << mSession->soap()->lastError();
+        }
+    }
+    mSession->forgetSession();
+}
+
+void SugarSoapProtocol::getEntriesCount(const ListEntriesScope &scope, const QString &moduleName, const QString &query)
+{
+    mSession->soap()->asyncGet_entries_count(mSession->sessionId(), moduleName, query, scope.deleted());
+}
+
+void SugarSoapProtocol::listEntries(const ListEntriesScope &scope, const QString &moduleName, const QString &query, const QString &orderBy, const QStringList &selectedFields)
+{
+    const int offset = scope.offset();
+    const int maxResults = 100;
+    const int fetchDeleted = scope.deleted();
+
+    KDSoapGenerated::TNS__Select_fields Fields;
+    Fields.setItems(selectedFields);
+
+    mSession->soap()->asyncGet_entry_list(mSession->sessionId(), moduleName, query, orderBy, offset, Fields, maxResults, fetchDeleted);
```

Comment:
same here ... but let's do that later. Let's sort out getEntriesCount first (including error handling), before we look at this one.